functions.py

```diff
@@ -0,0 +1,181 @@
+        for response, datum in enumerate(line.split(';')):
+            result[columns[response]] = datum.strip('"\n\r\t')
+        results.append(result)
+
+    return results
+
+# function makes the actual call to the api, phrase based reports
+# for the phrase based reports the request is done per keyword and per database (each geo has another one)
+# the result of the function is a df
+# as protection the function checks first if we didn't already updated today the tables
+def get_phrase_data(dimension, call_type, features, service_url, api_key, date_mid_month, kw_list, db_list, current_time, last_update_date):
+    if current_time.date() != last_update_date:
+        list_container = []
+        for keyword in range(0, len(kw_list)):
+            kw = kw_list[keyword]
+            for database in range(0, len(db_list)):
+                db = db_list[database]
+                response = request(dimension=dimension, call_type=call_type, api_key=api_key, geo=db,
+                                   features=features, service_url=service_url, phrase=kw,
+                                   date_mid_month=date_mid_month)
+                parsed_response = parse_response(call_data=response.content)
+
+                df = pd.json_normalize(parsed_response)
+                df['keyword'] = kw
+                df['database'] = db
+                list_container.append(df)
+
+        df_final = pd.concat(list_container, ignore_index=True)
+        df_final['display_date'] = datetime.strptime(date_mid_month, '%Y%m%d').date()
+        df_final['time_updated'] = current_time.strftime("%Y-%m-%d, %H:%M:%S")
+        return df_final
+    else:
+        print('Table is already updated')
+
+# function makes the actual call to the api, url based reports
+# for the url based reports the request is done per url (no separation to databases)
+# the result of the function is a df
+# as protection the function checks first if we didn't already updated today the tables
+def get_url_data(dimension, call_type, features, service_url, api_key, url_list, current_time, last_update_date):
+    if current_time.date() != last_update_date:
+        list_container = []
+        for url in range(0, len(url_list)):
+            url = url_list[url]
+            response = request(dimension=dimension, call_type=call_type, api_key=api_key, url=url,
+                               features=features, service_url=service_url)
+            parsed_response = parse_response(call_data=response.content)
+
+            df = pd.json_normalize(parsed_response)
+            list_container.append(df)
+
+        df_final = pd.concat(list_container, ignore_index=True)
+        df_final['display_date'] = df_final.apply(lambda row: uncode_utc(row['last_seen']), axis=1)
+        df_final['time_updated'] = current_time.strftime("%Y-%m-%d, %H:%M:%S")
+        return df_final
+    else:
+        print('Table is already updated')
+
+# execute presto sql scripts
+def execute_sql(sql_script, pc):

```

Comment:
according to the next comments , you won't need the methods execute_sql and get_quix_data